azkaban-spi/src/main/java/com/linkedin/azkaban/spi/Storage.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+
+package com.linkedin.azkaban.spi;
+
+import java.io.InputStream;
+import java.net.URI;
+
+
+/**
+ * The Azkaban Storage interface would facilitate getting and putting objects into a storage mechanism of choice.
+ * By default, this is set to the MySQL database. However, users can have the ability to choose between multiple
+ * storage types in future.
+ *
+ * This is different from storing Azkaban state in MySQL which would typically be maintained in a different database.
+ *
+ * Note: This is a synchronous interface.
+ */
+public interface Storage {
+
+  /**
+   * Get an InputStream object by providing a key.
+   *
+   * @param key The key is a URI pointing to the blob in Storage.
+   * @return InputStream for fetching the blob.
+   *         Returns null if the object doesn't exist.
+   */
+  InputStream get(URI key);
+
+  /**
+   * Put an object into Storage against a key. If the key already exists, then it throws
+   * {@link KeyAlreadyExistsException;}.
+   *
+   * @param key The key is a URI pointing to the blob in Storage.
+   * @param is The input stream from which the value is read to the store. The value is read completely
+   */
+  void put(URI key, InputStream is) throws KeyAlreadyExistsException;
```

Comment:
@chengren311 

Thanks for the review. Both work. 

My thought process of having an exception is to force the user (of this interface) to check for the key upfront before using it. A return value can be easily ignored and the user won't notice that a storage request failed. This may create problems in our case where user is unware that an artifact upload failed. By having a `RuntimeException` this is bubbled and the user is forced to handle this.

This is different from a Java Map API where it is Ok to just replace the value and return the current one.

Regarding get and delete, I updated the API to throw exceptions as to keep it consistent.